src/librustc/session/mod.rs

```diff
@@ -600,11 +600,51 @@ impl Session {
             .panic
             .unwrap_or(self.target.target.options.panic_strategy)
     }
-    pub fn linker_flavor(&self) -> LinkerFlavor {
+
+    fn linker(&self) -> &str {
         self.opts
-            .debugging_opts
-            .linker_flavor
-            .unwrap_or(self.target.target.linker_flavor)
+            .cg
+            .linker
+            .as_ref()
+            .and_then(|l| l.file_name())
+            .and_then(|l| l.to_str())
+            .unwrap_or_else(|| {
+                self.target.target.options.linker.as_ref().map(|l| &**l).unwrap_or_else(|| {
+                    "cc"
+                })
+            })
+    }
+
+    pub fn linker_flavor(&self) -> LinkerFlavor {
```

Comment:
This is the main change but the logic probably doesn't work correctly on Windows because of the .exe suffix.